### PR TITLE
Basic functionalities implementation

### PR DESCRIPTION
**Pull request by FelonEkonom**

### CODE REVIEW

File: .github/workflows/fetch_changes.yml [hunk 1]
```diff
@@ -32,7 +32,7 @@ jobs:
       # Runs a set of commands using the runners shell
```
Comment (mickel8) [inline .github/workflows/fetch_changes.yml]: This file should be deleted

File: README.md [hunk 2]
```diff
@@ -1,35 +1,29 @@
...
-[![API Docs](https://img.shields.io/badge/api-docs-yellow.svg?style=flat)](https://hexdocs.pm/membrane_template_plugin)
-[![CircleCI](https://circleci.com/gh/membraneframework/membrane_template_plugin.svg?style=svg)](https://circleci.com/gh/membraneframework/membrane_template_plugin)
+[![Hex.pm](https://img.shields.io/hexpm/v/membrane_opentelemetry.svg)](https://hex.pm/packages/membrane_opentelemetry)
+[![API Docs](https://img.shields.io/badge/api-docs-yellow.svg?style=flat)](https://hexdocs.pm/membrane_opentelemetry)
+[![CircleCI](https://circleci.com/gh/membraneframework/membrane_opentelemetry.svg?style=svg)](https://circleci.com/gh/membraneframework/membrane_opentelemetry)
 
-This repository contains a template for new plugins.
-
-Check out different branches for other flavours of template.
+This repository contains Membrane wrappers on OpenTelemetry functions.
```
Comment (daniel-jodlos) [inline README.md]: Being a little more verbose wouldn't hurt. After reading this, I still don't really know what I'm looking at

```suggestion
This repository contains Membrane wrappers of OpenTelemetry functions.
```

File: README.md [hunk 3]
```diff
@@ -1,35 +1,29 @@
...
 ```
 
-## Usage
-
-TODO
-
 ## Copyright and License
 
-Copyright 2020, [Software Mansion](https://swmansion.com/?utm_source=git&utm_medium=readme&utm_campaign=membrane_template_plugin)
+Copyright 2020, [Software Mansion](https://swmansion.com/?utm_source=git&utm_medium=readme&utm_campaign=membrane_opentelemetry)
```
Comment (daniel-jodlos) [inline README.md]:
```suggestion
Copyright 2022, [Software Mansion](https://swmansion.com/?utm_source=git&utm_medium=readme&utm_campaign=membrane_opentelemetry)
```

File: lib/opentelemetry.ex [hunk 4]
```diff
@@ -0,0 +1,213 @@
+defmodule Membrane.OpenTelemetry do
+  @moduledoc """
+  Defines macros for operations on OpenTelemetry spans.
+  Provided macros evalueate to appropriate calls to OpenTelemetry functions or to nothing, depending on config values
```
Comment (daniel-jodlos) [inline lib/opentelemetry.ex]:
```suggestion
  Provided macros expand to appropriate calls to OpenTelemetry functions or to nothing, depending on config values
```

File: lib/opentelemetry.ex [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+  @type span_name :: String.t()
+
+  @doc """
+  Starts a new span. If the second argument contains value `parent` under key `:parent`, then span named `parent` will be the parent span of the newly created one.
+  """
+  defmacro start_span(name, opts \\ []) do
+    if enabled(),
+      do: do_start_span(name, opts),
+      else: default_macro([name, opts])
```
Comment (daniel-jodlos) [inline lib/opentelemetry.ex]: do you need to call this? Why not just...
```suggestion
```

Also, `default_macro` is a little strange. It's effectively "no operation", right? So maybe let's call it `no_action` or something like that

File: lib/opentelemetry.ex [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    if enabled(),
+      do: do_add_events(name, events),
+      else: default_macro([name, events])
+  end
+
+  @doc """
+  Ensures, that every span started in the process calling this function, will be implicite closed after the process end.
+  Should be called in every process, that will execute any other function or macro from this module.
+  """
+  defmacro register() do
```
Comment (mat-hek) [inline lib/opentelemetry.ex]:
```suggestion
  defmacro register_process() do
```

File: lib/opentelemetry.ex [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+  require OpenTelemetry
+  require OpenTelemetry.Tracer
+  require OpenTelemetry.Ctx
+
+  @enabled Application.compile_env(:membrane_opentelemetry, :enabled, false)
+
+  @type span_name :: String.t()
+
+  @doc """
+  Starts a new span. If the second argument contains value `parent` under key `:parent`, then span named `parent` will be the parent span of the newly created one.
```
Comment (mat-hek) [inline lib/opentelemetry.ex]:
```suggestion
  Starts a new span. 
  
  Options:
  - `parent` - a span that will be the parent of the created span
```

File: lib/opentelemetry.ex [hunk 8]
```diff
@@ -0,0 +1,213 @@
+defmodule Membrane.OpenTelemetry do
+  @moduledoc """
+  Defines macros for operations on OpenTelemetry spans.
```
Comment (mat-hek) [inline lib/opentelemetry.ex]: Add better description and usage example

File: lib/opentelemetry.ex [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+      links =
+        case opts_map do
+          %{linked_spans: spans} when is_list(spans) -> spans
+          _else -> []
+        end
+        |> Enum.map(&OpenTelemetry.link/1)
+
+      new_span = OpenTelemetry.Tracer.start_span(unquote(name), %{links: links})
+      unquote(__MODULE__).ETSUtils.store_span(unquote(name), new_span)
+      OpenTelemetry.Tracer.set_current_span(new_span)
```
Comment (mat-hek) [inline lib/opentelemetry.ex]: Don't alter `current span` whenever possible

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/opentelemetry.ex (hunk 5 above)
Comment (mickel8) [reply lib/opentelemetry.ex]: If `default_macro` is needed, maybe `void` would be a siutable name?

**Review by mickel8 — APPROVED** [review]

File: README.md [hunk 10]
```diff
@@ -20,6 +20,71 @@ def deps do
...
+
+Before starting a span in your code, you have to call 
+```elixir
+Membrane.OpenTelemetry.register_process()
+```
+in process, that will open your spans. This function provides certainty, that if the process, that opened some spans, ends before closing them (eg. because of error), all these spans will be closed right after that.
+
+To start your fist span, you have to call 
+```elixir
+Membrane.OpenTelemetry.start_span("root_span")
```
Comment (mat-hek) [inline README.md:47]: I'd mention that "root_span" is the span ID and it has to be unique and what it's for

File: README.md [hunk 11]
```diff
@@ -1,35 +1,94 @@
...
   ]
 end
 ```
 
 ## Usage
 
-TODO
+Firstly, you have to turn on `Membrane.OpenTelemetry` in your config files. To do this, add 
+```elixir
+config :membrane_opentelemetry, enabled: telemetry_enabled
```
Comment (mat-hek) [inline README.md]:
```suggestion
config :membrane_opentelemetry, enabled: true
```

File: README.md [hunk 12]
```diff
@@ -1,35 +1,94 @@
...
+
+You can also specify span parent, by passing its `span_ctx`. It is useful, in the case, when you want to start a span, that will be a child of a span, that comes from another process and, for example, was passed to your process by message.
+```elixir
+Membrane.OpenTelemetry.start_span("another_child_span", parent_span: parent_span_ctx) 
+```
+In this case, you could also call 
+```elixir
+root_span_ctx = Membrane.OpenTelemetry.get_span("root_span")
+Membrane.OpenTelemetry.start_span("another_child_span", parent_span: root_span_ctx) 
+```
```
Comment (mat-hek) [inline README.md]:
````suggestion
To get the span context, call 
```elixir
root_span_ctx = Membrane.OpenTelemetry.get_span("root_span") 
```
in the process that started the `root_span`.
````

File: README.md [hunk 13]
```diff
@@ -1,35 +1,94 @@
...
+    otel_batch_processor: %{
+      exporter: {:otel_exporter_stdout, []}
+    }
+  ]
+```
+to your `config/runtime.exs` file. In this case, you will use exporter, that will put tracing data on stdout, but you can use Zipkin or Honeycomb exporters for OpenTelemetry as well (take a look on (membrane_videoroom)[https://github.com/membraneframework/membrane_videoroom] (config/runtime.exs)[https://github.com/membraneframework/membrane_videoroom/blob/master/config/runtime.exs], if you want an example).
+
+Before starting a span in your code, you have to call 
+```elixir
+Membrane.OpenTelemetry.register_process()
```
Comment (mat-hek) [inline README.md]: As discussed, let's do that automatically when starting the first span

File: README.md [hunk 14]
```diff
@@ -1,35 +1,94 @@
...
+In this case, you could also call 
+```elixir
+root_span_ctx = Membrane.OpenTelemetry.get_span("root_span")
+Membrane.OpenTelemetry.start_span("another_child_span", parent_span: root_span_ctx) 
+```
+
+If you want to have a span, that has a different id than a name, pass `:name` option to `Membrane.OpenTelemetry.start_span/2`
+```elixir
+Membrane.OpenTelemetry.start_span("grandchild_span", parent_id: "child_span", name: "grandchild_span_name") 
+```
```
Comment (mat-hek) [inline README.md]: Describe what's the name and how it differs from the id

File: lib/opentelemetry.ex [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+  @type span_id :: String.t()
+
+  @doc """
+  Starts a new span.
+
+  Options:
+  - `parent_span` - a span_ctx of a span, that will be the parent of the created span
+  - `parent_id` - an id of a span, that will be the parent of the created span. Shouldn't be used with `parent_span` option. Available only for spans, that were created or stored in the same process in which we call this macro.
+  - `name` - name of a span. If not provided, by default span name will be span id
+  """
```
Comment (mat-hek) [inline lib/opentelemetry.ex:25]: Write what it returns

**Review by mat-hek — CHANGES_REQUESTED** [review]

**Review by mat-hek — APPROVED** [review]

**Pull request merged**